src/dataflow/internal/process/functions/source.ts

```diff
@@ -0,0 +1,68 @@
+import type {RArgument, RParseRequestProvider} from '../../../../r-bridge'
+import { sourcedDeterministicCountingIdGenerator} from '../../../../r-bridge'
+import {requestProviderFromFile} from '../../../../r-bridge'
+import {type NormalizedAst, type ParentInformation, removeTokenMapQuotationMarks, type RFunctionCall, RType} from '../../../../r-bridge'
+import {RShellExecutor} from '../../../../r-bridge/shell-executor'
+import {executeSingleSubStep} from '../../../../core'
+import {type DataflowProcessorInformation, processDataflowFor} from '../../../processor'
+import {type DataflowScopeName, type Identifier, overwriteEnvironments, type REnvironmentInformation, resolveByName} from '../../../environments'
+import type {DataflowInformation} from '../../info'
+import {dataflowLogger} from '../../../index'
+
+let sourceProvider = requestProviderFromFile()
+
+export function setSourceProvider(provider: RParseRequestProvider): void {
+	sourceProvider = provider
+}
+
+export function isSourceCall(name: Identifier, scope: DataflowScopeName, environments: REnvironmentInformation): boolean {
+	if(name != 'source')
+		return false
+	const definitions = resolveByName(name, scope, environments)
+	return definitions !== undefined && definitions.some(d => d.kind == 'built-in-function')
+}
+
+export function processSourceCall<OtherInfo>(functionCall: RFunctionCall<OtherInfo & ParentInformation>, data: DataflowProcessorInformation<OtherInfo & ParentInformation>, information: DataflowInformation): DataflowInformation {
+	const sourceFile = functionCall.arguments[0] as RArgument<ParentInformation> | undefined
+	if(sourceFile?.value?.type == RType.String) {
+		const executor = new RShellExecutor()
+		const path = removeTokenMapQuotationMarks(sourceFile.lexeme)
+		const request = sourceProvider.createRequest(path)
+
+		// check if the sourced file has already been dataflow analyzed, and if so, skip it
+		const requestString = JSON.stringify(request)
+		if(data.referenceChain.some(r => JSON.stringify(r) == requestString)) {
```

Comment:
given the repeated use -> please at least outsource a fingerprinting function. 
Furthermore, the check appears to be rather expensive for deeper reference chains.